applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h

```diff
@@ -152,6 +152,105 @@ class KRATOS_API(SHAPE_OPTIMIZATION_APPLICATION) OptimizationUtilities
+/**
+     * Assemble a list of Numbers into a Vector, independent of the model part
+     */
+    static void AssembleBufferVector( Vector& rVector,
+        const std::vector<double>& rVariables)
+    {
+    	size_t VectorSize = rVariables.size();
+        if (rVector.size() != VectorSize){
+            rVector.resize(VectorSize);
+        }
+
+
+        for (size_t i=0;i < VectorSize;i++)
+        {
+            rVector[i] = rVariables[i];
+        }
+    }
+    /**
+     * Assemble a list of Numbers into a diagonal Matrix, independent of the model part
+     */
+    static void AssembleBufferMatrix( Matrix& rMatrix,
+        const std::vector<double>& rVariables)
+    {
+    	size_t VectorSize = rVariables.size();
+        rMatrix = ZeroMatrix(VectorSize, VectorSize);
+        for (size_t i=0; i < VectorSize; i++)
+        {
+            rMatrix(i,i) = rVariables[i];
+        }
+    }
+    /**
+     * Assemble a list of Vectors into a  Matrix
+     */
+    static void AssembleVectorstoMatrix(ModelPart& rModelPart,
+        Matrix& rMatrix,
+        const std::vector<Vector*>& rVariables
+    )
+    {
+        if ((rMatrix.size1() != rModelPart.NumberOfNodes()*3 || rMatrix.size2() !=  rVariables.size())){
+            rMatrix.resize(rModelPart.NumberOfNodes()*3, rVariables.size());
+        }
+
+
+    	int j=0;
+    	for (Vector* p_variable_j : rVariables)
+    	{
+        	const Vector& r_variable_j = *p_variable_j;
+
+
+        	for (size_t i = 0; i < rModelPart.NumberOfNodes()*3; i++)
+        	{
+        		rMatrix(i, j) = r_variable_j[i];
+        	}
+   		++j;
+ 	}
+
+    }
+
+    /**
+     * Calculate the relaxed projection of the objective gradient into the subspace tangent to
+     * the active constraint gradients.
+     * In a second step, calculate correction move
+     */
+    static void CalculateRelaxedProjectedSearchDirectionAndCorrection(
+        Vector& rObjectiveGradient,
+        Matrix& rConstraintGradients,
+        Matrix& rRelaxationCoefficients,
+        Vector& rCorrectionCoefficients,
+        LinearSolver<DenseSpace, DenseSpace>& rSolver,
+        Vector& rProjectedSearchDirection,
+        Vector& rCorrection
+        )
+    {
+        // local variable naming according to https://msulaiman.org/onewebmedia/GradProj_2.pdf
+        Vector& nabla_f = rObjectiveGradient;
+        Matrix& N = rConstraintGradients;
+        Vector& s = rProjectedSearchDirection;
+        Vector& c = rCorrection;
+        Matrix& omega_r = rRelaxationCoefficients;
+        Vector& omega_c = rCorrectionCoefficients;
+
+
+        Matrix NTN = prod(trans(N), N);
+        Matrix I = IdentityMatrix(N.size2());
+        Matrix NTN_inv(NTN.size1(), NTN.size2());
+
+        rSolver.Solve(NTN, NTN_inv, I); // solve with identity to get the inverse
+
+
+        s = - (nabla_f - prod(N, Vector(prod(omega_r, Vector(prod(NTN_inv, Vector(prod(trans(N), nabla_f))))))));
```

Comment:
```suggestion
        s = - (nabla_f - prod(N, prod(omega_r, prod(NTN_inv, prod(trans(N), nabla_f)))));
```

I would remove the vector conversion. It can be handled by the operators automatically.